juniper/src/executor/look_ahead_lazy.rs

```diff
@@ -0,0 +1,1555 @@
+        self.children.iter()
+    }
+}
+
+impl<'a, S: ScalarValue> IntoIterator for LookAheadChildren<'a, S> {
+    type Item = LookAheadSelection<'a, S>;
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.children.into_iter()
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub(super) enum SelectionSource<'a, S: ScalarValue> {
+    Field(&'a Field<'a, S>),
+    Spread {
+        field_name: &'a str,
+        set: Option<&'a [Selection<'a, S>]>,
+    },
+}
+
+/// A selection performed by a query
+#[derive(Clone, Copy, Debug)]
+pub struct LookAheadSelection<'a, S: ScalarValue + 'a> {
+    source: SelectionSource<'a, S>,
+    applies_for: Applies<'a>,
+    vars: &'a Variables<S>,
+    fragments: &'a HashMap<&'a str, Fragment<'a, S>>,
+}
+
+impl<'a, S: ScalarValue> LookAheadSelection<'a, S> {
+    pub(super) fn new(
+        source: SelectionSource<'a, S>,
+        vars: &'a Variables<S>,
+        fragments: &'a HashMap<&'a str, Fragment<'a, S>>,
+    ) -> Self {
+        Self {
+            source,
+            applies_for: Applies::All,
+            vars,
+            fragments,
+        }
+    }
+
+    /// Returns the original name of the field, represented by the current selection.
+    pub fn field_original_name(&self) -> &'a str {
+        match self.source {
+            SelectionSource::Field(field) => field.name.item,
+            SelectionSource::Spread { field_name, .. } => field_name,
+        }
+    }
+
+    /// Returns the alias of the field, represented by the current selection, if any.
+    pub fn field_alias(&self) -> Option<&'a str> {
+        match self.source {
+            SelectionSource::Field(field) => field.alias.map(|alias| alias.item),
+            SelectionSource::Spread { .. } => None,
+        }
+    }
+
+    /// Returns the (potentially aliased) name of the field, represented by the current selection.
+    pub fn field_name(&self) -> &'a str {
+        self.field_alias()
+            .unwrap_or_else(|| self.field_original_name())
+    }
+
+    /// Returns the top level arguments from the current selection.
+    pub fn arguments(&self) -> Option<LookAheadArguments<'a, S>> {
+        match self.source {
+            SelectionSource::Field(field) => {
+                field
+                    .arguments
+                    .as_ref()
+                    .map(|spanned_arguments| LookAheadArguments {
+                        arguments: &spanned_arguments.item,
+                        vars: self.vars,
+                    })
+            }
+            _ => None,
+        }
+    }
+
+    /// Returns the children from the current selection.
+    pub fn children(&self) -> LookAheadChildren<'a, S> {
+        self.build_children(Applies::All)
+    }
+
+    /// Returns the children from the current selection that only applies to a specific type.
+    pub fn children_for_explicit_type(&self, type_name: &str) -> LookAheadChildren<'a, S> {
```

Comment:
This is the replacement for the `ConcreteLookAheadSelection` struct. I didn't fully understand why that exists as a separate type, this simplifies things IMO.

The largest difference between the old and new code is that the type filter is not "persistent" anymore. But I think filtering fields for a specific type is most meaningful at a specific document level anyway.

Reply:
`fn children()` is an eager computation that collects into a vector. That's why I made this alternative API `children_for_explicit_type`: It also collects into a vector, only with a stricter filter.

`children().for_explicit_type("Person")` would allocate a vector and then filter that into a new vector again, so I would consider this API less "Rusty" in the sense that it doesn't encourage the cheapest possible way to perform some computation.

Reply:
@audunhalland 

> edit: That said, even `fn children()` could _theorectically_ return an iterator of lazily computed children without any vector allocation, but since there is no generator/coroutine in Rust yet I considered the implementation too complex.

Looked into details, and agree that it would be too complex for the sake of decoupled methods.

Let's go with `children_for_explicit_type()` then, and leave a `// TODO:` comment in hope for generators stabilization.

Reply:
Thinking about it again, `fn children()` would always require allocation in Rust, since it involves potentially unbounded recursion (visiting FragmentSpreads that could refer to other FragmentSpreads, is that a valid model in GraphQL)?

The way it's done how involves recursion using the stack, and pushing the results to the end of a vector. A Rust generator would be stackless, turning things around: To support recursion it would need to use the heap for _that_, while yielding the actual items would be lazy. We'd be forced to use heap allocation in either case.